dali/python/nvidia/dali/_debug_mode.py

```diff
@@ -419,57 +446,122 @@ def _external_source(self, name=None, **kwargs):
-            to_external_source, _, data = _PipelineDebug._classify_data(value)
-            if to_external_source != kwargs_es[key]:
-                raise RuntimeError(
-                    f"In operator '{op_name}' argument '{key}' {unexpected_argument_msg(to_external_source)}.")
-            if to_external_source:
-                pipe.feed_input(f'{self._es_kwarg_name}{key}', data)
+            is_batch, device, data = _PipelineDebug._classify_data(value)
 
-        res = pipe.run()
-        res = tuple([DataNodeDebug(tensor, **data_node.__dict__)
-                    for tensor, data_node in zip(res, pipe._graph_out)])
+            check_batch_classification(
+                kwargs_classification[key].is_batch, is_batch, 'Argument', key)
+            check_device_classification(kwargs_classification[key].device, device, 'Argument', key)
 
-        self._debug_on = True
+            if not is_batch and data != init_args[key]:
+                raise RuntimeError(f"Argument '{key}' for operator '{op_name}' unexpectedly changed"
+                                   f" value from '{init_args[key]}' to '{data}'")
+            if is_batch:
+                call_args[key] = data
+
+        input_sets = op_helper._prep_input_sets(inputs)
+        res = [self._run_op_on_device(op_name, op_device, input, call_args) for input in input_sets]
+
+        if len(res) == 1:
+            res = self._pack_to_data_node_debug(res[0], op_name)
+        else:
+            res = op_helper._repack_output_sets(res)
+            res = self._pack_to_data_node_debug(res, op_name)
 
         if len(res) == 1:
             return res[0]
 
         return res
 
-    def _wrap_op_call(self, op_wrapper, inputs, kwargs):
-        self._cur_subpipeline_id += 1
-        key = inspect.getframeinfo(
-            inspect.currentframe().f_back.f_back)[:3] + (self._cur_subpipeline_id,)
-        if not self._subpipelines_built:
-            self._subpipelines[key] = self._create_subpipeline(op_wrapper, inputs, kwargs)
+    @staticmethod
+    def _extract_data_node_inputs(_, *inputs):
+        """Extracts DataNodeDebugs from inputs for arithmetic operator and transforms data to GPU if needed."""
+        data_nodes = []
+        to_gpu = any([input.device == 'gpu' for input in inputs if isinstance(input, DataNodeDebug)])
+
+        for input in inputs:
+            if isinstance(input, DataNodeDebug):
+                if to_gpu and input.device != 'gpu':
+                    data_nodes.append(input.gpu())
+                else:
+                    data_nodes.append(input)
+
+        return data_nodes
 
-        if key in self._subpipelines:
-            return self._run_subpipeline(self._subpipelines[key], op_wrapper.__name__, inputs, kwargs)
+    def _wrap_op_call(self, op_name, inputs, kwargs):
+        self._cur_op_id += 1
+        key = inspect.getframeinfo(
+            inspect.currentframe().f_back.f_back)[:3] + (self._cur_op_id,)
+        if not self._ops_built:
```

Comment:
Maybe it is just only me, but having module called `_ops` and class property `_ops` leads to confusion.

Reply:
fixed